File: shared/web-sdk/src/connectors/delegation-progress-connector.ts
import type { DelegationStatus } from "@sentient/protocol";
import type { Connector, SentientSDKInternal } from "../connector-types.ts";
import { createLogger } from "../logger.ts";

const log = createLogger(["sentient", "sdk", "connectors", "delegation-progress"]);

export interface DelegationProgressItem {
  /** Handle returned by the background tool. Dedup key. */
  readonly taskId: string;
  /** The turn that dispatched the delegation. */
  readonly turnId: string;
  /** Worker the task was delegated to (e.g. "hermes"). */
  readonly agent: string;
  readonly status: DelegationStatus;
  /** Optional short progress note from the worker. */
  readonly note?: string;
}

export interface DelegationProgressConnectorConfig {
  /** Called on every individual update. */
  onUpdate?: (item: DelegationProgressItem) => void;
  /** Called whenever the list changes — insertion order (dispatch order). */
  onList?: (items: readonly DelegationProgressItem[]) => void;
}

// ---------------------------------------------------------------------------
// DelegationProgressConnector — live status of BACKGROUND delegated work
// (spec §5.4 / §7).
//
// Capability: "delegation.progress"
// Direction: status (observer)
//
// Receives: delegation.progress — one frame per status transition of a
// `delegateTask` background task; dedup by taskId so a row transitions in
// place. The frame carries no timestamp, so ordering is insertion (dispatch)
// order. Terminal rows stay in the list; the UI filters if it wants only
// running work.
//
// Join key: `ToolCallSnapshotItem.taskId` (from turn.tool.update) is the same
// id, so the UI can attach progress to the tool pill that launched it.
//
// CACHE LIFETIME — attach/detach are TRANSPORT lifecycle, not session
// lifecycle, so neither touches the task map. Same rule, same reason, as
// ToolStatusConnector's own CACHE LIFETIME header and
// ConversationHistoryConnector's mirror: a reconnect detaches every connector
// (sdk-close-handler.teardownWsForReconnect) and re-attaches them on the next
// session.ready, and the `recovered:true` resume then replays ONLY the frames
// the client missed. Clearing here dropped every row the client had already
// seen, and the gateway never re-sends those.
//
// The stakes are higher on this connector than on the live tool cache.
// Delegated work is the longest-lived thing on this wire — `delegateTask` is
// fire-and-steer, so a Hermes task routinely outlives the socket that
// dispatched it and reports its completion on a LATER turn. Wiping the map on
// reconnect stranded exactly that work: its `running` row vanished, and the
// only frame that could have restored it was the terminal one, arriving
// minutes later, if at all.
//
// `reset()` is the session/identity teardown path (SentientSDK.disconnect) and
// the only clear. A reconnect must never reach it.
// ---------------------------------------------------------------------------

export class DelegationProgressConnector implements Connector {
  readonly capability = "delegation.progress";
  readonly kind = "status" as const;

  private readonly config: DelegationProgressConnectorConfig;
  private unsubs: (() => void)[] = [];
  private tasks = new Map<string, DelegationProgressItem>();

  constructor(config: DelegationProgressConnectorConfig = {}) {
    this.config = config;
  }

  /** Every delegated task seen this session, in dispatch order. */
  list(): readonly DelegationProgressItem[] {
    return [...this.tasks.values()];
  }

  attach(sdk: SentientSDKInternal): void {
    this.unsubs.push(
      sdk.onMessage("delegation.progress", (msg: unknown) => {
        const m = msg as {
          taskId?: string;
          turnId?: string;
          agent?: string;
          status?: DelegationStatus;
          note?: string;
        };
        if (!m.taskId || !m.turnId || !m.agent || !m.status) {
          log.warn("delegation-progress-dropped", {
            reason: "missing required field on delegation.progress",
            taskId: m.taskId ?? null,
          });
          return;
        }
        const item: DelegationProgressItem = {
          taskId: m.taskId,
          turnId: m.turnId,
          agent: m.agent,
          status: m.status,
          ...(m.note !== undefined ? { note: m.note } : {}),
        };
        this.tasks.set(item.taskId, item);
        log.info("delegation-progress", {
          taskId: item.taskId,
          turnId: item.turnId,
          agent: item.agent,
          status: item.status,
        });
        this.config.onUpdate?.(item);
        this.config.onList?.(this.list());
      }),
    );
  }

  /** Drop the socket subscriptions only. The task map survives — see CACHE
   *  LIFETIME above. */
  detach(): void {
    for (const unsub of this.unsubs) unsub();
    this.unsubs = [];
  }

  /** Drop all session state. `SentientSDK.disconnect()` is the one caller. */
  reset(): void {
    const previousCount = this.tasks.size;
    this.tasks = new Map();
    log.info("delegation-progress.reset", { reason: "session identity teardown", previousCount });
  }
}

File: shared/mobile-sdk/src/commonMain/kotlin/io/sentient/mobilesdk/connectors/DelegationProgressConnector.kt
// ---------------------------------------------------------------------------
// DelegationProgressConnector — live progress for BACKGROUND delegateTask work
// (design §5.4 / §7). Closest analog: the retired TaskStatusConnector (same
// upsert-by-id shape; superseded by TaskListConnector's full-state model),
// but a delegation row is agent-scoped and outlives the turn that dispatched it —
// its completion returns later as a stimulus, so terminal rows stay in the list and
// the UI filters if it wants.
//
// PRIVACY: `note` is worker-authored text (it can quote user content). Log its LENGTH
// only, never the note.
//
// Threading: single-threaded; the router drives handle() on the orchestrator's
// dispatcher. The mutable map is owned here.
// ---------------------------------------------------------------------------
package io.sentient.mobilesdk.connectors

import io.sentient.mobilesdk.log.createLogger
import io.sentient.mobilesdk.protocol.SdkEvent
import io.sentient.mobilesdk.protocol.ServerMessage

/** Immutable snapshot of one delegated background task. */
data class DelegationSnapshotItem(
    val taskId: String,
    /** The turn that dispatched the delegation (join key for the UI's turn grouping). */
    val turnId: String,
    /** Worker identity, e.g. "hermes". */
    val agent: String,
    /** "running" | "done" | "error". */
    val status: String,
    /** Short worker-authored progress note; null when absent. USER-ADJACENT — never logged. */
    val note: String? = null,
)

class DelegationProgressConnector(
    private val onList: ((List<DelegationSnapshotItem>) -> Unit)? = null,
    private val onEvent: ((SdkEvent) -> Unit)? = null,
) : Connector {
    override val capability: String = CAPABILITY

    private val log = createLogger("connector", "delegation-progress")

    private val tasks = LinkedHashMap<String, DelegationSnapshotItem>()

    /** All delegation rows in arrival order. Terminal rows are retained. */
    fun list(): List<DelegationSnapshotItem> = tasks.values.toList()

    override fun handle(msg: ServerMessage) {
        when (msg) {
            is ServerMessage.DelegationProgress -> onProgress(msg)
            else -> Unit // not owned by this connector
        }
    }

    private fun onProgress(msg: ServerMessage.DelegationProgress) {
        val item = DelegationSnapshotItem(
            taskId = msg.taskId,
            turnId = msg.turnId,
            agent = msg.agent,
            status = msg.status,
            note = msg.note,
        )
        log.info(
            "delegation.progress",
            mapOf(
                "taskId" to item.taskId,
                "turnId" to item.turnId,
                "agent" to item.agent,
                "status" to item.status,
                "noteLen" to (item.note?.length ?: 0),
            ),
        )
        tasks[item.taskId] = item
        onEvent?.invoke(SdkEvent.DelegationProgressed(item))
        onList?.invoke(list())
    }

    /**
     * Drop every row. Driven by the orchestrator's `clearConversationScopedState` from
     * EVERY leave-the-conversation entry point — the awaited `newChat` / `switchSession`
     * and the fire-and-forget `sendNewChat` / `sendSwitchSession` the UI actually uses.
     * Never driven by interrupt or by the reconnect re-establish: a delegation outlives
     * its turn and survives a resume of the same conversation.
     */
    fun clear() {
        if (tasks.isEmpty()) return
        log.info("clear", mapOf("count" to tasks.size))
        tasks.clear()
        onList?.invoke(list())
    }

    companion object {
        const val CAPABILITY: String = "delegation.progress"
    }
}
